### ISSUE TITLE
Nested control styles

### ISSUE THREAD

**Issue by HariboTer** [nifty-gui/nifty-gui#105]
When nested controls come in combination with substyles, they either do not work as expected or I am misunderstanding it.
The matter is that I have this controlDefinition (for clarity I'll focus on the important parts):

``` xml
<controlDefinition name="intField" defaultFocusElement="textField" controller="controls.IntField">
    <panel childLayout="horizontal">
      <control id="textField" name="textfield" style="#textField" />
    </panel>
</controlDefinition>
```

So I use that controlDefinition in my main .xml like

``` xml
<control name="intField" id="someId" style="my-intfield" />
```

The corresponding style xml is:

``` xml
<style id="my-intfield#textField" base="my-textfield" />
```

And for completeness, the my-textfield:

``` xml
<nifty-styles xmlns="http://nifty-gui.sourceforge.net/nifty-styles-1.3.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://nifty-gui.sourceforge.net/nifty-styles-1.3.xsd http://nifty-gui.sourceforge.net/nifty-styles-1.3.xsd">
  <registerMouseCursor id="textFieldCursor" filename="main/data/interface/nifty/my-style/textfield/mouse-cursor-textfield.png" hotspotX="3" hotspotY="12"/>

  <!-- the background of the textfield -->
  <style id="my-textfield#panel">
    <attributes childLayout="overlay" height="23px" />
    <effect>
      <onHover name="changeMouseCursor" id="textFieldCursor" />
      <onHover name="border" color="#822f" post="true" />
    </effect>
  </style>

  <!-- the actual input field -->
  <style id="my-textfield#field">
    <attributes childLayout="center" childClip="true" backgroundColor="#666f" padding="0px,2px"/>
    <effect>
      <onActive name="border" color="#222f" post="true" inset="1px" />
      <onFocus name="colorBar" color="#080f" post="true" inset="1px" />
      <onEnabled name="renderQuad" startColor="#2228" endColor="#2220" post="true" length="150" />
      <onDisabled name="renderQuad" startColor="#2220" endColor="#2228" post="true" length="150" />
    </effect>
  </style>

  <!-- the text in the input field -->
  <style id="my-textfield#text" base="base-font">
    <attributes color="#000f" selectionColor="#f00f" visibleToMouse="false" align="left" valign="center" textHAlign="left" />
    <effect>
      <onFocus name="textColor" post="false" color="#cccf" />
    </effect>
  </style>

  <!-- the cursor is rendered in a separate layer on top of the input field and this is the parent panel of this -->
  <style id="my-textfield#cursor-panel">
    <attributes childLayout="absolute" padding="0px,2px" />
  </style>

  <!-- the actual cursor -->
  <style id="my-textfield#cursor">
    <attributes filename="main/data/interface/nifty/my-style/textfield/cursor-empty.png" />
    <effect>
      <onCustom name="imageOverlayPulsate" period="250" timeType="infinite" pulsateType="rectangle" filename="main/data/interface/nifty/my-style/textfield/cursor.png" post="true"/>
    </effect>
  </style>

</nifty-styles>
```

my-textfield is a valid style, at least I think so as I can use it without nesting and it works nicely. But in this context, an exception is thrown by nifty.
The problem is:
I would expect Nifty to look up the style my-intfield#textField. Then it would request the parts of it that the textfield control needs, so it would go deeper and look up for my-intfield#textField#field etc, which would be given (I think) as my-intfield#textField uses my-textfield as base. But actually Nifty doesn't! It only keeps one style and one substyle in mind, so instead of looking for my-intfield#textField#field, it looks for my-intfield#field, which of course it doesn't find and voila, exception and byebye cruel world.
My workaround for now is to point the control directly to its style in the control definition, so it is

``` xml
<controlDefinition name="intField" defaultFocusElement="textField" controller="controls.IntField">
    <panel childLayout="horizontal">
      <control id="textField" name="textfield" style="my-textfield" />
    </panel>
</controlDefinition>
```

instead of

``` xml
<controlDefinition name="intField" defaultFocusElement="textField" controller="controls.IntField">
    <panel childLayout="horizontal">
      <control id="textField" name="textfield" style="#textField" />
    </panel>
</controlDefinition>
```

, but this doesn't seem as nice to me as the nested styles would. If this is a desired behaviour of Nifty, I would like to know the reason behind that. Else, I hope I could point in the right direction for a fix ;)

**Comment by HariboTer** [comment]
Unfortunately this problem just gained more cruciality to me, because some nifty controls (e. g. the scrollPanel) uses internal nifty controls (e. g. scrollbars). The style of the controls can be set "externally" - those of their corresponding sub controls not. So to use an own style on them, you would have to go to e. g. the original scrollPanel xml file in the nifty-default-controls.jar file, extract it, edit e. g. their scrollbars' style and put it back in. Ugly, but at least it works. But now for me the problem arised that I need two different styles for scrollpanels. Which I can set for them - but not for their scrollbars! I also tried something like

``` xml
<controlDefinition someControlStyle="default-style">
  <control name="someControl" style="$someControlStyle" />
</controlDefinition>
```

, but this causes some internal error in the NiftyGUI library, as it throws an exception when this code snippet would come to use, even although the style seems to have been assigned properly (there is no exception when I assign the style directly though).
A response about this would be much appreciated!

**Comment by void256** [comment]
Yes, this is currently a somewhat annoying missing feature. The last time I tried to fix it was not successfully since the whole style and sub-style business is kinda, erm, bad in Nifty :) I'm not at all happy with it and will probably take another look but for the moment this is something that just doesn't work unfortunately (That's why its mentioned in the offical Nifty manual as well).

The workaround you described (changing the original style definition) is correct. You could even include your own style definition to get that working I think, so that it's not necessary to change the original jar file.

And you're right with your second point as well: If you need more than one different style than you are currently kinda doomed. The only thing I can think of would be to workaround this with plain old brute force: copy the complete control definition and use different names like "scrollPanelRed" and "scrollPanelGreen" or something. So that you can actually change the "source" of the control to point to another style. Not very nice but might work.

**Comment by HariboTer** [comment]
Thanks for your confirmation :-) The workaround you explained works, but sadly only partially, because the dropDown-control uses, for its dropdown menu, a popup with a fixed id, which logically can't be defined twice with respectively different styles :-S

**Comment by HariboTer** [comment]
After all, I went ahead, copied all required sources from the DropDownControl in my own controls package and modified them to accept the popupId as a parameter. That way it finally works ... but oh dear xD

**Comment by ghost** [comment]
I just got bit by something similar. Apparently, if you have an xml control definition file, and you specify a style in the `<controldefinition>` tag, and then you have another `<control>` in that same control definition, you can't specify a substyle for that control. It will just be ignored. You can only specify a "normal" style that doesn't use the `#`.

The archetypal example for this issue is `nifty-console.xml` in the `nifty-controls` package. Let's take a look:

``` xml
<?xml version="1.0" encoding="UTF-8"?>
<nifty-controls xmlns="http://nifty-gui.lessvoid.com/nifty-gui">
    <controlDefinition name="nifty-console" controller="de.lessvoid.nifty.controls.console.ConsoleControl">
        <panel style="nifty-console-panel">
            <control id="#listBox" name="listBox" style="nifty-console-listbox" displayItems="$lines"
                     selectionMode="Disabled" horizontal="off" vertical="optional">
                <text text="Template" style="nifty-console-listbox-item"
                      controller="de.lessvoid.nifty.controls.listbox.ListBoxItemController"
                      inputMapping="de.lessvoid.nifty.input.mapping.MenuInputMapping"/>
            </control>
            <control id="#textInput" name="textfield" style="nifty-console-textfield"/>
        </panel>
    </controlDefinition>
</nifty-controls>
```

Do you notice anything strange about the above xml? Yes - it's not using substyles. It looks like we could make it more "nifty" by converting it to use an overall style file with substyles...

So, we convert it, and... it doesn't work...?!

Notice how the following **_should**_ work for control substyles (assuming that `nifty-console-style.xml` is correctly written of course), but does not. It does work for non-controls, however, such as panel's, text, & images:

``` xml
<?xml version="1.0" encoding="UTF-8"?>
<nifty-controls xmlns="http://nifty-gui.lessvoid.com/nifty-gui">
    <controlDefinition name="nifty-console" style="nifty-console-style" controller="de.lessvoid.nifty.controls.console.ConsoleControl">
        <panel style="nifty-console-style#panel"> <!-- This panel substyle will work -->

           <!-- This listbox control substyle will be ignored -->
            <control id="#listBox" name="listBox" style="nifty-console-style#listbox" displayItems="$lines"
                     selectionMode="Disabled" horizontal="off" vertical="optional">

                <!-- This text substyle "should" work, but who knows, since it's inside a listBox control... -->
                <text text="Template" style="nifty-console-style#listbox-item"
                      controller="de.lessvoid.nifty.controls.listbox.ListBoxItemController"
                      inputMapping="de.lessvoid.nifty.input.mapping.MenuInputMapping"/>
            </control>

           <!-- This textfield control substyle will be ignored -->
            <control id="#textInput" name="textfield" style="nifty-console-style#textfield"/>
        </panel>
    </controlDefinition>
</nifty-controls>
```

If you create a custom control set up like the second example, good luck debugging. You won't get any errors or warnings. Your control substyles will simply be ignored.

**Comment by ghost** [comment]
@void256 The use case I specified above seems much less complex than @HariboTer's. It seems like something that would not be too difficult to fix. I had a difficult time, however, understanding Nifty's style resolving code.

**Comment by relu91** [comment]
I hope that @void256 has time to make a little light on style thing. Everytime I've look at it, I couldn't find how it won't work... mostly styles work but if you are not using controls, that is the 99% of use cases. Also I've found that setStyle doesn't work for controls, but I not sure and I need to test it again.

**Comment by void256** [comment]
Style resolving is, unfortunately, one of the worst parts in Nifty. The last time I tried, I gave up. It's a difficult beast to battle. I'm not sure what to do with it, to be honest :/

**Comment by ghost** [comment]
Someone needs to write lots of tests for it and then create some tests that fail, such as substyles on controls, and then refactor to make the tests pass. Not saying it's going to be easy. But we need to do something, not just give up. @void256 If you could explain HOW the style resolving code works, with snippets from the actual codebase, it might help. Also if you could explain where you're getting stuck. I wouldn't mind attempting some refactoring but I don't really understand where to start. I looked at `StyleResolver`, `StyleResolverControlDefinition`, `StyleResolverDefault`, and their usages, but I can't figure out what's going on.

**Comment by ghost** [comment]
The lack of errors & warnings is the most disturbing part. Failing silently is the worst kind of evil. I would not wish it upon my enemies.

**Comment by void256** [comment]
OK! Maybe we should review it together on IRC or something? I don't wanna go to that dark place alone again :sweat_smile: :satisfied: We should arrange a meeting!

**Comment by ghost** [comment]
Sure!

**Comment by ghost** [comment]
@mkaring Would love to have your help with this, since you seem to know Nifty better than anyone.

**Comment by mkaring** [comment]
@3xp0n3nt I'll join in. But that layout resolving is nasty (not nifty). Its not going to be pretty.
@void256 Give me a ping on gtalk in case I do not react in IRC.

**Comment by void256** [comment]
@mkaring Nice! Will do!

@3xp0n3nt I know you usually work at cruel times :wink: but I think it's best you give us some timeframe or possible times for a meeting that will still be comfortable for you considering your time zone :smile: The plan is to take a first look at that style resolving mess and I'll try to remember what the hell I was thinking when writing it :wink: and then we'll see what we can do about it ... print it and burn it or something .. hehe

**Comment by relu91** [comment]
Can you guys make a summary of your discussion here? :smiley: I would love to listen how the style system will change, but I'm pretty busy now , so probably I won't be on IRC when you will discuss about that :P .

Thanks

**Comment by ghost** [comment]
@void256 @mkaring How about the usual IRC meeting time? 10 PM GMT? I
believe that's the correct time.

Aaron Mahan
Co-Founder & CEO
Forerunner Games, LLC
http://forerunnergames.com/
Peril: http://perilgame.com/

**Comment by void256** [comment]
Yes, that's a good idea! @3xp0n3nt  What day would be best for you?

PS: Sorry, was busy this week .. real life got me again but I now have a new bathroom :wink:

**Comment by ghost** [comment]
@void256 You're bathroom is being remodeled too? o_O

**Comment by ghost** [comment]
@void256 I'll have to get back to you next week about a time, I'm swamped at the moment.

**Comment by Genhis** [comment]
Has there been any progress with the issue? It is very difficult to make complex nested controls without nested styles.

**Comment by Genhis** [comment]
A workaround for this issue is to remove an IF statement in `de.lessvoid.nifty.loaderv2.types.ElementType#applyStyleInternal()`.

So if anyone wanted to have nested styles, change
```
if (!style.startsWith("#")) {
      return new StyleResolverControlDefinintion(styleResolver, style);
}
```
to
```
return new StyleResolverControlDefinintion(styleResolver, style);
```

It would require to rewrite most of the XML style definitions because all of the styles would become nested. Moreover, I'm not sure whether it has any side effects but so far, it works fine for me.
